python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,223 @@
+          or keys_as_set.issubset(
+              {"value", "variance", "unit", "dtype", "shape", "dims"})):
+        # Case of a Variable-like dict (most-likely)
+        return _dict_to_variable(dict_obj)
+    else:
+        # Case of a Dataset-like dict
+        out = sc.Dataset()
+        for key, item in dict_obj.items():
+            out[key] = _dict_to_data_array(item)
+        return out
+
+
+def _dict_to_variable(d):
+    """
+    Convert a python dict to a scipp Variable.
+    """
+    # The Variable constructor does not accept both `shape` and `values`. If
+    # `values` is present, remove `shape` from the list. Also remove `dims` in
+    # the case of a 0D variable.
+    keylist = list(d.keys())
+    if "value" in keylist:
+        if "shape" in keylist:
+            keylist.remove("shape")
+        if "dims" in keylist:
+            keylist.remove("dims")
+    if "values" in keylist and "shape" in keylist:
+        keylist.remove("shape")
+    out = {}
+
+    is_event_data = False
+    if "dtype" in keylist:
+        is_event_data = str(d["dtype"]).startswith("event_list_float")
+
+    # TODO: maybe this constructor would be worth adding to the scipp module
+    # itself?
+    if is_event_data:
+        if "shape" not in d:
+            shp = d["values"].shape
+        else:
+            shp = d["shape"]
+
+        var = sc.Variable(dims=d["dims"],
+                          shape=shp,
+                          dtype=getattr(sc.dtype, str(d["dtype"])))
+
+        ndim = len(d["dims"])
+        indices = tuple()
+        for i in range(ndim):
+            indices += range(shp[i]),
+        # Now construct all indices combinations using itertools
+        for ind in product(*indices):
+            # And for each indices combination, slice the original data
+            vslice = var
+            aslice = d["values"]
+            for i in range(ndim):
+                vslice = vslice[d["dims"][i], ind[i]]
+                aslice = aslice[ind[i]]
+            vslice.values = aslice
+        return var
+
+    else:
+        for key in keylist:
+            if key == "dtype" and isinstance(d[key], str):
+                out[key] = getattr(sc.dtype, d[key])
+            else:
+                out[key] = d[key]
+        return sc.Variable(**out)
+
+
+def _dict_to_data_array(d):
+    """
+    Convert a python dict to a scipp DataArray.
+    """
+    if ("data" not in d) and ("unaligned" not in d):
+        raise KeyError("To create a DataArray, the supplied dict must contain "
+                       "either 'data' or 'unaligned'. "
+                       "Got {}.".format(d.keys()))
+    out = {"coords": {}, "masks": {}, "attrs": {}}
+    for key in out.keys():
+        if key in d:
+            for name, item in d[key].items():
+                out[key][name] = _dict_to_variable(item)
+    if "unaligned" in d:
+        out["unaligned"] = _dict_to_data_array(d["unaligned"])
+    else:
+        out["data"] = _dict_to_variable(d["data"])
```

Comment:
I think think you can move the variable directly. Is there a reason you aren' already doing this?

`out["data"] = sc.detail.move(_dict_to_variable(d["data"]))`